### PR TITLE
Deprecate legacy URI handling & return valid ASCII-safe URIs

### LINKED ISSUE #27475: Deprecate legacy/invalid URI parsing in Project#file / Project#uri
These allow passing invalid URIs to get a valid result back, but we should reject these malformed URIs up-front in order to avoid odd parsing issues like #26678 or mishandling of UTF-8 encoded codepoints.

### PR DESCRIPTION
**Pull request by octylFractal**
Fixes #27475 

This attempts to provide consistency between the inputs accepted and the outputs emitted by Gradle in terms of URIs, and to be stricter about what can be provided as input to prevent ambiguity about how certain URIs should be handled.

### CODE REVIEW

File: platforms/software/resources/src/main/java/org/gradle/internal/resource/ExternalResourceName.java [hunk 1]
```diff
@@ -94,9 +94,9 @@ public URI getUri() {
      */
     public String getDecoded() {
         if (encodedRoot == null) {
-            return path;
+            return encode(path, false);
```
Comment (jvandort) [inline platforms/software/resources/src/main/java/org/gradle/internal/resource/ExternalResourceName.java:97]: Can you explain why this change was made?
It seems very odd that `getDecoded` returns the result of `encode`

**Review by jvandort — CHANGES_REQUESTED** [review]
I'd like to better understand the changes to `AbstractResourcePattern.java` and `ExternalResourceName.java`

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 2]
```diff
@@ -35,6 +35,33 @@ The previous step will help you identify potential problems by issuing deprecati
...
+== Upgrade from 8.6 and earlier
+
+=== Potential breaking changes
+
+=== Deprecations
+
+[[deprecated_legacy_url_decoding]]
+==== Deprecated legacy URL decoding behavior
+Prior to Gradle 8.3, Gradle would decode a `CharSequence` given to `link:{groovyDslPath}/org.gradle.api.Project.html#org.gradle.api.Project:uri(java.lang.Object)[Project.uri(Object)]`
+using a flawed algorithm that accepted invalid URLs and improperly decoded others. Gradle now uses the `URI` class to parse and decode URLs,
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:48]:
```suggestion
Gradle now uses the `URI` class to parse and decode URLs, but with a fallback to the legacy behavior in the event of an error.
```

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 3]
```diff
@@ -35,6 +35,33 @@ The previous step will help you identify potential problems by issuing deprecati
...
+[[changes_8.7]]
+== Upgrade from 8.6 and earlier
+
+=== Potential breaking changes
+
+=== Deprecations
+
+[[deprecated_legacy_url_decoding]]
+==== Deprecated legacy URL decoding behavior
+Prior to Gradle 8.3, Gradle would decode a `CharSequence` given to `link:{groovyDslPath}/org.gradle.api.Project.html#org.gradle.api.Project:uri(java.lang.Object)[Project.uri(Object)]`
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:47]:
```suggestion
Prior to Gradle 8.3, Gradle would decode a `CharSequence` given to `link:{groovyDslPath}/org.gradle.api.Project.html#org.gradle.api.Project:uri(java.lang.Object)[Project.uri(Object)]` using an algorithm that accepted invalid URLs and improperly decoded others. 
```

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 4]
```diff
@@ -35,6 +35,33 @@ The previous step will help you identify potential problems by issuing deprecati
...
+using a flawed algorithm that accepted invalid URLs and improperly decoded others. Gradle now uses the `URI` class to parse and decode URLs,
+but with a fallback to the legacy behavior in the event of an error.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:49]:
```suggestion
```

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 5]
```diff
@@ -35,6 +35,33 @@ The previous step will help you identify potential problems by issuing deprecati
...
+
+Starting in Gradle 9.0, the fallback will be removed, and the error will be thrown instead.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:51]:
```suggestion
Starting in Gradle 9.0, the fallback will be removed, and an error will be thrown instead.
```

**Review by lkasso — APPROVED** [review]
Minor doc changes.

File: platforms/software/resources/src/main/java/org/gradle/internal/resource/ExternalResourceName.java (hunk 1 above)
Comment (octylFractal) [reply platforms/software/resources/src/main/java/org/gradle/internal/resource/ExternalResourceName.java:97]: I think "decoded" was a bad name for this method, as it was already half-encoded (the root). I changed the name to match its usage.

File: platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/FileOrUriNotationConverter.java [hunk 6]
```diff
@@ -84,17 +85,33 @@ public void convert(Object notation, NotationConvertResult<? super Object> resul
...
+                    // Bad file URI, return URI as-is
+                }
             }
+            result.converted(uri);
             return;
         }
         if (notation instanceof CharSequence) {
             String notationString = notation.toString();
             if (notationString.startsWith("file:")) {
-                result.converted(new File(uriDecode(notationString.substring(5))));
```
[the remark this answers is not in the document]
Comment (octylFractal) [reply platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/FileOrUriNotationConverter.java:97]: Yes, this was a merge fail.

**Comment by octylFractal** [comment]
@bot-gradle test PT

File: platforms/documentation/docs/src/snippets/dependencyManagement/customizingResolution-conditionalSubstitutionRule/groovy/buildSrc/src/main/groovy/myproject.java-library-conventions.gradle [hunk 7]
```diff
@@ -3,5 +3,5 @@ plugins {
 }
 
 repositories {
-    ivy { url "file://${rootProject.projectDir}/repo" }
+    ivy { url "${rootProject.projectDir.toURI().toASCIIString()}/repo" }
```
Comment (big-guy) [inline platforms/documentation/docs/src/snippets/dependencyManagement/customizingResolution-conditionalSubstitutionRule/groovy/buildSrc/src/main/groovy/myproject.java-library-conventions.gradle:6]: Does this work?

url = uri(rootProject.layout.projectDirectory.dir("repo"))

**Review by big-guy — CHANGES_REQUESTED** [review]

File: platforms/extensibility/plugin-use/src/integTest/groovy/org/gradle/plugin/repository/RepositoryOrderingIntegrationSpec.groovy [hunk 8]
```diff
@@ -61,7 +61,7 @@ class RepositoryOrderingIntegrationSpec extends AbstractIntegrationSpec {
         when:
         settingsFile << """
             pluginManagement {
-                repositories { maven { url = "$pluginManagementRepoUri" } }
+                repositories { maven { url = "pluginManagementRepo" } }
```
[the remark this answers is not in the document]
Comment (octylFractal) [reply platforms/extensibility/plugin-use/src/integTest/groovy/org/gradle/plugin/repository/RepositoryOrderingIntegrationSpec.groovy:64]: Not sure, I think I fixed the root cause later. I'll revert this.

File: platforms/documentation/docs/src/snippets/dependencyManagement/customizingResolution-conditionalSubstitutionRule/groovy/buildSrc/src/main/groovy/myproject.java-library-conventions.gradle (hunk 7 above)
Comment (octylFractal) [reply platforms/documentation/docs/src/snippets/dependencyManagement/customizingResolution-conditionalSubstitutionRule/groovy/buildSrc/src/main/groovy/myproject.java-library-conventions.gradle:6]: Yes. Changed.

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ArtifactDependenciesIntegrationTest.groovy [hunk 9]
```diff
@@ -244,15 +244,15 @@ task listMissingClassifier {
 
         failure.assertHasCause("""Could not find lib-1.0.zip (org.gradle.test:lib:1.0).
 Searched in the following locations:
-    ${module.artifactFile(type: 'zip').toURL()}""")
+    ${module.artifactFile(type: 'zip').toURI().toASCIIString()}""")
```
[the remark this answers is not in the document]
Comment (octylFractal) [reply platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ArtifactDependenciesIntegrationTest.groovy:247]: Yes, added.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/AbstractResourcePattern.java [hunk 10]
```diff
@@ -59,7 +59,9 @@ private AbstractResourcePattern(ExternalResourceName pattern) {
 
     @Override
     public String getPattern() {
-        return pattern.getDecoded();
+        // Replace encoded [] with plain variants, since we are using them with a special meaning here.
+        // Inefficient but easy to read, I don't think this method is hot.
+        return pattern.getDisplayable().replace("%5B", "[").replace("%5D", "]");
```
[the remark this answers is not in the document]
Comment (octylFractal) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/AbstractResourcePattern.java:64]: The encoded `[]` come from the normal URI encoding rules, which state that certain characters are reserved for having special meaning in schemes and must be encoded; see `gen-delims` in [RFC 3986 Section 2.2](https://datatracker.ietf.org/doc/html/rfc3986#section-2.2). Since in the pattern, these actually do have a special meaning, I decode them here so that patterns are not displayed with encoded `[]`.

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 11]
```diff
@@ -35,6 +35,32 @@ The previous step will help you identify potential problems by issuing deprecati
...
+To fix a deprecation warning, invalid URLs that require the legacy behavior should be re-encoded to be valid URLs, such as in the following examples:
+
+.Legacy URL Conversions
+|===
+| Original Input | New Input | Reasoning
+
+| `file:relative/path` | `relative/path` | The `file` scheme does not support relative paths.
+| `file:relative/path%21` | `relative/path!` | Without a scheme, the path is taken as-is, without decoding.
+| `https://example.com/my folder/` | `https://example.com/my%20folder/` | Spaces are not valid in URLs.
+| `https://example.com/my%%badly%encoded%path` | `https://example.com/my%25%25badly%25encoded%25path` | `%` must be encoded as `%25` in URLs, and no `%`-escapes should be invalid.
```
[the remark this answers is not in the document]
Comment (octylFractal) [reply platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:61]: No, I don't think so. This applies to `project.file`, `UrlArtifactRepository#setUrl(Object)`, and more.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/AbstractResourcePattern.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (octylFractal) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/AbstractResourcePattern.java:64]: I benchmarked this against the loop I would've written, and it turns out that the `replace` calls are actually faster!
```
Benchmark                                                                                                               (rawUri)  Mode  Cnt    Score    Error  Units
UriPerf.decodeUriReplace                            https://foo.bar/repo/[organization]/[revision]/[module](.[classifier]).[ext]  avgt   10  127.136 ±  8.467  ns/op
UriPerf.decodeUriReplace      https://foo.bar/repo/[organization]/[revision]/[module]/[artifact]-[revision](.[classifier]).[ext]  avgt   10  165.374 ± 16.068  ns/op
UriPerf.decodeUriReplaceLoop                        https://foo.bar/repo/[organization]/[revision]/[module](.[classifier]).[ext]  avgt   10  216.316 ±  9.639  ns/op
UriPerf.decodeUriReplaceLoop  https://foo.bar/repo/[organization]/[revision]/[module]/[artifact]-[revision](.[classifier]).[ext]  avgt   10  281.815 ±  7.584  ns/op
```
So I'll remove the efficiency comment, as I assumed incorrectly.

File: platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/repositories/resolver/IvyResourcePatternTest.groovy [hunk 12]
```diff
@@ -37,7 +37,7 @@ class IvyResourcePatternTest extends Specification {
         "http://host/"     | ""                 | "http://host/"
         "http://host"      | ""                 | "http://host"
         "http://host/"     | "/"                | "http://host/"
-        "http://host/repo" | "query?[revision]" | "http://host/repo/query?[revision]"
+        "http://host/repo" | "query?[revision]" | "http://host/repo/query%3F[revision]"
```
[the remark this answers is not in the document]
Comment (octylFractal) [reply platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/repositories/resolver/IvyResourcePatternTest.groovy:40]: I only changed the display of the name here, not any of the actual logic. I can think of two options:
1. Make this display raw `?` again, to reflect how it's used.
2. Make a follow-up PR to split the pattern on `?`, and pass it to `ExternalResourceName` properly.

What do you think?

[the remark this answers is not in the document]
Comment (big-guy) [reply platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/repositories/resolver/IvyResourcePatternTest.groovy:40]: SGTM

**Review by big-guy — APPROVED** [review]

File: platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/repositories/resolver/IvyResourcePatternTest.groovy (hunk 12 above)
[the remark this answers is not in the document]
Comment (octylFractal) [reply platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/repositories/resolver/IvyResourcePatternTest.groovy:40]: Choosing to do neither option as upon review the implementation actually does not interpret `?` as the query parameter, so the display of `?` as encoded is correct here.

**Comment by octylFractal** [comment]
@bot-gradle test and merge

**Review by bamboo — APPROVED** [review]
LGTM

**Comment by octylFractal** [comment]
@bot-gradle test and merge
